### PR TITLE
screen-space refraction

### PR DESCRIPTION
**Pull request by pixelflinger**
With this PR screen-space refraction is functional. 

Caveats:

The blur passes for supporting rough refractive materials is quite 
heavy and increases with the resolution.

SSR uses a gaussian approximation for the brdf and therefore doesn’t 
match perfectly cubemap-based refraction and IBL.

The use of MSAA with screen-space effects, while working, is going
to incur a large cost, especially on tilers.

ssr

### CODE REVIEW

File: filament/src/fg/FrameGraph.cpp [hunk 1]
```diff
@@ -455,11 +455,22 @@ FrameGraph& FrameGraph::compile() noexcept {
...
             for (fg::RenderTarget& rt : renderTargets) {
                 auto& textures = rt.desc.attachments.textures;
-                if (textures[0].isValid()) {
-                    FrameGraphHandle handle = textures[0];
+                FrameGraphHandle handle = textures[0];
+                if (handle.isValid()) {
                     ResourceNode const& node = resourceNodes[handle.index];
                     if (node.resource->imported && node.resource == from.resource) {
                         for (size_t i = 1; i < textures.size(); ++i) {
+                            // FIXME: here we're likely leaking the resources we're clearing,
```
Comment (romainguy) [inline filament/src/fg/FrameGraph.cpp]: File a bug?

Comment (pixelflinger) [reply filament/src/fg/FrameGraph.cpp]: The "leak" is only during the frame, it's really not that bad --  I track those small issues with FIXMEs.

File: filament/src/PostProcessManager.cpp [hunk 2]
```diff
@@ -631,4 +685,140 @@ FrameGraphId<FrameGraphTexture> PostProcessManager::blurPass(FrameGraph& fg,
...
+
+                PostProcessMaterial const& separableGaussianBlur = mSeparableGaussianBlur;
+                FMaterialInstance* const mi = separableGaussianBlur.getMaterialInstance();
+
+                PipelineState pipeline;
+                pipeline.program = separableGaussianBlur.getProgram();
+                pipeline.rasterState = separableGaussianBlur.getMaterial()->getRasterState();
+                pipeline.scissor = mi->getScissor();
+
+                float2 kernel[128];
```
[the remark this answers is not in the document]
Comment (pixelflinger) [reply filament/src/PostProcessManager.cpp]: 128 coefficients, actually corresponds to a kernel width of 509  :-)

File: filament/src/RenderPass.h [hunk 3]
```diff
@@ -88,7 +88,8 @@ class RenderPass {
     enum class Pass : uint64_t {    // 6-bits max
         DEPTH    = uint64_t(0x00) << PASS_SHIFT,
         COLOR    = uint64_t(0x01) << PASS_SHIFT,
-        BLENDED  = uint64_t(0x02) << PASS_SHIFT,
+        REFRACT  = uint64_t(0x02) << PASS_SHIFT,
+        BLENDED  = uint64_t(0x03) << PASS_SHIFT,
```
[the remark this answers is not in the document]
Comment (pixelflinger) [reply filament/src/RenderPass.h]: It doesn't matter I think right now, they're treated as a value (but used for sorting), not as a bitfield.

**Pull request merged**